Do not serialize shorthands when any longhands are missing

darinadler opened the pull request:
#### a7f14bf80bc907a6b436de4c81853fc0359acabd
<pre>
Do not serialize shorthands when any longhands are missing
<a href="https://bugs.webkit.org/show_bug.cgi?id=249440">https://bugs.webkit.org/show_bug.cgi?id=249440</a>
rdar://problem/103423965

Reviewed by Tim Nguyen.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
Expect one more test to pass.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::commonShorthandChecks const): Return empty string when a longhand is
missing. Also redid the logic for values that are set by the shorthand, since it no longer needs
to compute a count.
(WebCore::StyleProperties::getPropertyValue const): Removed unneeded checks since
commonShorthandChecks handles missing longhands, and parsing makes sure things have correct types.
(WebCore::StyleProperties::offsetValue const): Ditto.
(WebCore::StyleProperties::fontVariantValue const): Ditto.
(WebCore::StyleProperties::get2Values const): Ditto.
(WebCore::StyleProperties::get4Values const): Ditto.
(WebCore::StyleProperties::getLayeredShorthandValue const): Ditto.
(WebCore::StyleProperties::getGridTemplateValue const): Ditto.
(WebCore::StyleProperties::getGridValue const): Ditto.
(WebCore::StyleProperties::getGridRowColumnShorthandValue const): Ditto.
(WebCore::StyleProperties::getGridAreaShorthandValue const): Ditto.
(WebCore::StyleProperties::getShorthandValue const): Ditto.
(WebCore::StyleProperties::getCommonValue const): Ditto.
(WebCore::StyleProperties::borderImagePropertyValue const): Ditto.
(WebCore::StyleProperties::borderRadiusShorthandValue const): Ditto.
(WebCore::StyleProperties::borderPropertyValue const): Ditto.

* Source/WebCore/css/StyleProperties.h: Removed the unneeded argument from fontVariantValue.

Canonical link: <a href="https://commits.webkit.org/258038@main">https://commits.webkit.org/258038@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d750deffd9e22b8e713fb44d6aa78abd3a818d43

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-16-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-BigSur-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 🧪 win ](https://ews-build.webkit.org/#/builders/Windows-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/WinCairo-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-BigSur-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-BigSur-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-16-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-BigSur-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Ventura-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-9-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-BigSur-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 jsc-mips ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5514 "Found 1 new test failure: fast/scrolling/arrow-key-scroll-in-rtl-document.html (failure)") | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-9-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🧪 jsc-mips-tests ](https://ews-build.webkit.org/#/builders/JSC-MIPSEL-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2884 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -502,84 +497,39 @@ String StyleProperties::offsetValue() const
-    if (offsetPositionIndex != -1) {
-        auto offsetPosition = propertyAt(offsetPositionIndex);
-        if (!offsetPosition.isImplicit()) {
-            if (!offsetPosition.value())
```
I'm assuming it couldn't really happen that a property which isn't missing had a null value?

darinadler replied:
That's right. No such thing.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -593,12 +543,8 @@ String StyleProperties::fontVariantValue(const StylePropertyShorthand& shorthand
-
         // Bypass getPropertyCSSValue which will return an empty string for system keywords.
-        auto value = propertyAt(foundPropertyIndex).value();
+        auto value = propertyAt(findPropertyIndex(property)).value();
```
I think this could just use `getPropertyCSSValue`. You would need to check for null, but it would replace the `isSystemFontShorthand` check below.

darinadler replied:
Good idea. That would be better.

darinadler replied:
Done.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -647,20 +593,10 @@ String StyleProperties::fontSynthesisValue() const
-    if (!start.value() || !end.value())
-        return { };
+    auto start = propertyAt(findPropertyIndex(shorthand.properties()[0]));
+    auto end = propertyAt(findPropertyIndex(shorthand.properties()[1]));
```
`end` could just use `getPropertyCSSValue`.

`start` has an `isImplicit` check which doesn't seem to make much sense, maybe it could use `getPropertyCSSValue` too.

darinadler replied:
The `isImplicit` check is going away soon.

I think I‘ll take a pass through tightening up the code again once I finish removing implicit entirely.

Loirooriol commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -151,14 +151,9 @@ String StyleProperties::commonShorthandChecks(const StylePropertyShorthand& shor
     std::optional<bool> importance;
     size_t numSetFromShorthand = 0;
```
`numSetFromShorthand` kinda becomes useless. Only need to track if the 1st longhand had `hasBeenSetFromShorthand`, and if there is a mismatch immediately return `emptyString()`.

nt1m reviewed (approved):
Great cleanup!

Would be nice to clean up `propertyAt(findPropertyIndex(X)` vs. `getPropertyCSSValue(X)`.

darinadler committed:
Do not serialize shorthands when any longhands are missing
https://bugs.webkit.org/show_bug.cgi?id=249440
rdar://problem/103423965

Reviewed by Tim Nguyen.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
Expect one more test to pass.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::commonShorthandChecks const): Return empty string when a longhand is
missing. Also redid the logic for values that are set by the shorthand, since it no longer needs
to compute a count.
(WebCore::StyleProperties::getPropertyValue const): Removed unneeded checks since
commonShorthandChecks handles missing longhands, and parsing makes sure things have correct types.
(WebCore::StyleProperties::offsetValue const): Ditto.
(WebCore::StyleProperties::fontVariantValue const): Ditto.
(WebCore::StyleProperties::get2Values const): Ditto.
(WebCore::StyleProperties::get4Values const): Ditto.
(WebCore::StyleProperties::getLayeredShorthandValue const): Ditto.
(WebCore::StyleProperties::getGridTemplateValue const): Ditto.
(WebCore::StyleProperties::getGridValue const): Ditto.
(WebCore::StyleProperties::getGridRowColumnShorthandValue const): Ditto.
(WebCore::StyleProperties::getGridAreaShorthandValue const): Ditto.
(WebCore::StyleProperties::getShorthandValue const): Ditto.
(WebCore::StyleProperties::getCommonValue const): Ditto.
(WebCore::StyleProperties::borderImagePropertyValue const): Ditto.
(WebCore::StyleProperties::borderRadiusShorthandValue const): Ditto.
(WebCore::StyleProperties::borderPropertyValue const): Ditto.

* Source/WebCore/css/StyleProperties.h: Removed the unneeded argument from fontVariantValue.

Canonical link: https://commits.webkit.org/258038@main
